gptel.el

```diff
@@ -912,6 +912,60 @@ Later plists in the sequence take precedence over earlier ones."
         (setq rtn (plist-put rtn p v))))
     rtn))
 
+;; https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random)
+(defun gptel--uuid ()
+  "Generate a UUID v4-1."
+  (format "%04x%04x-%04x-4%03x-8%03x-%04x%04x%04x"
+          (random #x10000) (random #x10000)
+          (random #x10000)
+          (random #x1000)
+          (random #x1000)
+          (random #x10000) (random #x10000) (random #x10000)))
+
+(defun gptel--machine-id ()
+  "Generate a machine ID."
+  (let ((hex-chars "0123456789abcdef")
+        (length 65)
+        hex)
+    (dotimes (_ length)
+      (setq hex (nconc hex (list (aref hex-chars (random 16))))))
+    (apply #'string hex)))
+
+(cl-defun gptel--url-retrieve (url &key method data headers)
+  "Retrieve URL synchronously with METHOD, DATA and HEADERS."
+  (let ((url-request-method (if (eq method'post) "POST" "GET"))
+        (url-request-data (encode-coding-string (gptel--json-encode data) 'utf-8))
+        (url-mime-accept-string "application/json")
+        (url-request-extra-headers
+         `(("content-type" . "application/json")
+           ("editor-plugin-version" . "CopilotChat.nvim/2.0.0")
+           ("user-agent" . "CopilotChat.nvim/2.0.0")
+           ("editor-version" . "Neovim/0.10.0")
+           ,@headers)))
+    (with-current-buffer (url-retrieve-synchronously url)
+      (goto-char url-http-end-of-headers)
+      (gptel--json-read))))
+
+(defun gptel--restore (file)
+  "Restore saved object from FILE."
+  (when (file-exists-p file)
+    (car (read-from-string
+          (let ((coding-system-for-read 'utf-8-auto-dos))
```

Comment:
I'm confused about why the coding system for read is `utf-8-auto-dos` while the coding system for write below in `gptel--save` is `utf-8-unix`.

Since the token is alphanumeric, the difference is going to be a an extra `^M` char at the end, which probably won't matter.  But is there any reason these two are different?

Reply:
The reason of using `utf-8-auto-dos` for read is so that the `^M` character will be ignored. So for example, the save file is opened by some editor in Windows that automatically changing the EOL to CRLF, we can still properly read it.
Whereas on saving, we should save as `utf-8-unix` style so that no `^M` characters got in

Reply:
Please add a short comment with this explanation (1-line) where you set `utf-8-auto-dos`, otherwise I'll be confused again in the future.

Reply:
Done